File: python-examples/e-commerece-category/api/ecommerece-category.py
from playwright.async_api import Page, BrowserContext
from intuned_browser import go_to_url
from typing import List
from runtime_helpers import extend_payload
from urllib.parse import urlparse, urljoin
from utils.types_and_schemas import EcommereceCategoryParams, Category


async def handle_cookies(page: Page) -> None:
    """
    Handles cookie consent banner if present.
    Replace the selector with the appropriate one for your store.
    """
    try:
        # Replace selector with appropriate one for your store
        await page.wait_for_selector("#onetrust-accept-btn-handler", timeout=5000)
        await page.click("#onetrust-accept-btn-handler")
        print("Accepted cookies")
        await page.wait_for_timeout(1000)
    except Exception:
        pass


async def extract_categories(page: Page, store_url: str) -> List[Category]:
    """
    Extracts category links from the main menu.
    Replace selectors with appropriate ones for your store.
    """
    categories: List[Category] = []

    # Replace selector with appropriate one for your store's menu
    menu_selector = ".has-submenu a.main-menu__link"
    links = await page.query_selector_all(menu_selector)

    for link in links:
        href = await link.get_attribute("href")
        name = await link.inner_text()

        if href:
            # Make sure href is absolute URL
            category_url = urljoin(store_url, href)
            category_name = name.strip() if name else ""

            categories.append(
                {
                    "category_name": category_name,
                    "category_url": category_url,
                }
            )

    return categories


async def automation(
    page: Page,
    params: EcommereceCategoryParams,
    context: BrowserContext | None = None,
    **_kwargs,
):
    """
    Scrapes all category links from an e-commerce store's main menu.
    Each category will be scraped separately using the extend_payload mechanism.

    Example params:
    {
        "store_url": "https://www.example-store.com/"
    }

    This function collects category metadata (name, url) from the main menu
    and then uses extend_payload to trigger individual scrapes for each category's products.
    """
    await page.set_viewport_size({"width": 1280, "height": 800})
    params = EcommereceCategoryParams(**params)
    if not params.store_url:
        raise ValueError("store_url is required in params")

    store_url = params.store_url.rstrip("/")

    # Parse and validate the store URL
    parsed_url = urlparse(store_url)
    if not parsed_url.scheme or not parsed_url.netloc:
        raise ValueError(f"Invalid store URL: {store_url}")

    print(f"Scraping categories from: {store_url}")

    # Navigate to the store home page
    await go_to_url(page, store_url)

    # Handle cookie consent banner
    await handle_cookies(page)

    # Wait for the main menu to be present
    # Replace selector with appropriate one for your store
    await page.wait_for_selector(".has-submenu a.main-menu__link")

    # Extract all categories from the menu
    all_categories = await extract_categories(page, store_url)

    print(f"Total categories found: {len(all_categories)}")

    # Enqueue each category for further scraping
    for category in all_categories:
        extend_payload(
            {
                "api": "ecommerece-list",
                "parameters": dict(category),
            }
        )

    return {"categories": all_categories}

File: typescript-examples/e-commerece-category/api/ecommerece-category.ts
import { Page, BrowserContext } from "playwright";
import { extendPayload } from "@intuned/runtime";
import { goToUrl } from "@intuned/browser";
import { z } from "zod";
import { categorySchema, Category } from "../utils/typesAndSchemas.js";

type Params = z.infer<typeof categorySchema>;

/**
 * Handles cookie consent banner if present.
 * Replace the selector with the appropriate one for your store.
 */
async function handleCookies(page: Page): Promise<void> {
  try {
    // Replace selector with appropriate one for your store
    await page.waitForSelector("#onetrust-accept-btn-handler", {
      timeout: 5000,
    });
    await page.click("#onetrust-accept-btn-handler");
    console.log("Accepted cookies");
    await page.waitForTimeout(1000);
  } catch {
    // Cookie banner not present or already accepted
  }
}

/**
 * Extracts category links from the main menu.
 * Replace selectors with appropriate ones for your store.
 */
async function extractCategories(
  page: Page,
  storeUrl: string
): Promise<Category[]> {
  const categories: Category[] = [];

  // Replace selector with appropriate one for your store's menu
  const menuSelector = ".has-submenu a.main-menu__link";
  const links = await page.$$(menuSelector);

  for (const link of links) {
    const href = await link.getAttribute("href");
    const name = await link.innerText();

    if (href) {
      // Make sure href is absolute URL
      const categoryUrl = new URL(href, storeUrl).toString();
      const categoryName = name?.trim() || "";

      categories.push({
        category_name: categoryName,
        category_url: categoryUrl,
      });
    }
  }

  return categories;
}

/**
 * Scrapes all category links from an e-commerce store's main menu.
 * Each category will be scraped separately using the extendPayload mechanism.
 *
 * Example params:
 * {
 *   "store_url": "https://www.example-store.com/"
 * }
 *
 * This function collects category metadata (name, url) from the main menu
 * and then uses extendPayload to trigger individual scrapes for each category's products.
 */
async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
): Promise<{ categories: Category[] }> {
  await page.setViewportSize({ width: 1280, height: 800 });

  if (!params || !params.store_url) {
    throw new Error("store_url is required in params");
  }

  const storeUrl = params.store_url.replace(/\/$/, "");

  // Parse and validate the store URL
  try {
    new URL(storeUrl);
  } catch {
    throw new Error(`Invalid store URL: ${storeUrl}`);
  }

  console.log(`Scraping categories from: ${storeUrl}`);

  // Navigate to the store home page
  await goToUrl({ page, url: storeUrl });

  // Handle cookie consent banner
  await handleCookies(page);

  // Wait for the main menu to be present
  // Replace selector with appropriate one for your store
  await page.waitForSelector(".has-submenu a.main-menu__link");

  // Extract all categories from the menu
  const allCategories = await extractCategories(page, storeUrl);

  console.log(`Total categories found: ${allCategories.length}`);

  // Enqueue each category for further scraping
  for (const category of allCategories) {
    extendPayload({
      api: "ecommerece-list",
      parameters: category,
    });
  }

  return { categories: allCategories };
}

export default handler;
